curvesim/iterators/param_samplers/parameterized_pool_iterator.py

```diff
@@ -0,0 +1,166 @@
+            Pool parameters set before all simulations.
+
+        pool_map : dict, optional
+            See __new__ method.
+
+        """
+        self._validate_pool_type(pool)
+        self.pool_template = deepcopy(pool)
+        self.set_pool_attributes(self.pool_template, fixed_params)
+        self.parameter_sequence = self.make_parameter_sequence(variable_params)
+
+    def __iter__(self):
+        """
+        Yields
+        -------
+        pool : :class:`~curvesim.templates.SimPool`
+            A pool object with the current variable parameters set.
+
+        params : dict
+            A dictionary of the pool parameters set on this iteration.
+        """
+        for params in self.parameter_sequence:
+            pool = deepcopy(self.pool_template)
+            self.set_pool_attributes(pool, params)
+            yield pool, params
+
+    def make_parameter_sequence(self, variable_params):
+        """
+        Returns a list of dicts for each possible combination of the input parameters.
+
+        Parameters
+        ----------
+        variable_params: dict
+            Pool parameters to vary across simulations.
+
+            Keys: pool parameters, Values: iterable of values
+
+        Returns
+        -------
+        List(dict)
+            A list of dicts defining the parameters for each iteration.
+        """
+
+        keys, values = zip(*variable_params.items())
+        self._validate_attributes(self.pool_template, keys)
+
+        sequence = []
+        for vals in product(*values):
+            params = dict(zip(keys, vals))
+            sequence.append(params)
+
+        return sequence
+
+    def _validate_pool_type(self, pool):
+        """Validates that the input pool is an instance of self._pool_type."""
+        if not isinstance(pool, self._pool_type):
+            input_class = pool.__class__.__name__
+            expected_class = self._pool_type.__name__
+            self_class = self.__class__.__name__
+
+            raise ParameterSamplerError(
+                f"Parameter sampler '{self_class}' only supports pool type "
+                f"'{expected_class}'; recieved '{input_class}'."
+            )
+
+    @property
+    @abstractmethod
+    def _pool_type(self):
+        """The expected pool type for a ParameterizedPoolIterator subclass."""
+        raise NotImplementedError
+
+
+class ParameterizedCurvePoolIterator(CurvePoolMixin, ParameterizedPoolIterator):
+    """
+    :class:`ParameterizedPoolIterator` parameter sampler specialized for Curve pools.
+    """
+
+
+class ParameterizedCurveMetaPoolIterator(CurveMetaPoolMixin, ParameterizedPoolIterator):
+    """
+    :class:`ParameterizedPoolIterator` parameter sampler specialized for Curve meta-pools.
+    """
+
+
+class ParameterizedCurveCryptoPoolIterator(
+    CurveCryptoPoolMixin, ParameterizedPoolIterator
+):
+    """
+    :class:`ParameterizedPoolIterator` parameter sampler specialized for Curve crypto pools.
+    """
+
+
+DEFAULT_POOL_MAP = {
```

Comment:
Might as well make this part of the class so it's easier to grab and modify.

Reply:
This turned out to be a bit messier than expected because it introduces a circularity. DEFAULT_POOL_MAP contains subclasses of ParameterizedPoolIterator, which of course have to be defined after the parent class is defined, so they can't be referenced inside the parent class.

A few options, all imperfect:
- Use strings of the subclass names in the mapping, and refer to them in __new__ using globals()["SubclassName"] or getattr() on the module (issue: ugly/code less clear)
- Define the class level pool map in __new__, possibly letting it be set via the pool_map kwarg (issue: seems dangerous)
- add the pool map to the class at the bottom of the module (issue: ugly/code less clear)
- Do nothing (issue: no class level override; intended scope less clear)

The right answer probably depends on how important it is to set the mapping at the class level vs. overriding it just for a particular instance (which is already possible with the pool_map kwarg).